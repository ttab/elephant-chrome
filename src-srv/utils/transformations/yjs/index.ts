import * as Y from 'yjs'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import type { GetDocumentResponse } from '../../../protos/service.js'
import { type TextbitElement } from '@ttab/textbit'
import { revertNewsdoc } from '../newsdoc/index.js'

export function yjsStateAsUpdate (content: TextbitElement[], data: Y.Doc): Uint8Array {
  const insertContentDelta = slateNodesToInsertDelta(content)
  const sharedContentRoot = data.get('content', Y.XmlText) as Y.XmlText
  sharedContentRoot.applyDelta(insertContentDelta)
  data.share.set('content', sharedContentRoot as Y.AbstractType<any>)

  return Y.encodeStateAsUpdate(data)
}

/**
 * Helper function to transform a newsDoc to a Y.Map()
 * @param data GetDocumentResponse
 * @param map Y.Map<unknown>
 * @returns Y.Map<unknown>
 */
export function newsDocToYmap (data: GetDocumentResponse, map: Y.Map<unknown>): Y.Map<unknown> {
  const d = {
    ...data,
    version: data.version.toString()
  }

  for (const key in d) {
    const value = d[key as keyof typeof d]
    if (typeof value === 'object') {
      const m = new Y.Map()
      if (Array.isArray(value)) {
        const a = new Y.Array()
        if (value?.length > 0) {
          a.push([newsDocToYmap(value as unknown as GetDocumentResponse, m)])
        }
        map.set(key, a)
      } else {
        for (const k in value) {
          m.set(k, value[k as keyof typeof value])
        }
        map.set(key, m)
      }
    }
    map.set(key, value)
  }
  return map
}

/**
* Helper function to transform a Document(Y.Doc) to a newsDoc
* @param document Document
* @returns GetDocumentResponse
*/
export function yDocToNewsDoc (document: Y.Doc): GetDocumentResponse {
  const original = document.get('original', Y.Map) as Y.Map<unknown>
  const json = original.toJSON()

  // revert content
  const sharedRoot = document.get('content', Y.XmlText) as Y.XmlText
  const content = yTextToSlateElement(sharedRoot).children
  json.document.content = revertNewsdoc(content as any)
  json.version = BigInt(json.version)

  return json as GetDocumentResponse
}
