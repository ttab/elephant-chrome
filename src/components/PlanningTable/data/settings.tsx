import {
  SignalHigh,
  SignalMedium,
  SignalLow,
  FileType,
  Image,
  Camera,
  Video
} from '@ttab/elephant-ui/icons'

export const priorities = [
  {
    label: 'High',
    value: '1',
    icon: SignalHigh,
    color: '#FF5050'
  },
  {
    label: 'Medium',
    value: '2',
    icon: SignalMedium
  },
  {
    label: 'Low',
    value: '3',
    icon: SignalLow
  }
]

export const assignmentTypes = [
  {
    label: 'Text',
    value: 'text',
    icon: FileType
  },
  {
    label: 'Graphic',
    value: 'graphic',
    icon: Image
  },
  {
    label: 'Picture',
    value: 'picture',
    icon: Camera
  },
  {
    label: 'Video',
    value: 'video',
    icon: Video
  }
]
