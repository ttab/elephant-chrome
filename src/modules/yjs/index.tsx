export { WebSocketProvider } from './contexts/WebSocketProvider'
