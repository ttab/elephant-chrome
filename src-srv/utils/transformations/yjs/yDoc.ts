import * as Y from 'yjs'
import { type Block, type GetDocumentResponse } from '@/protos/service.js'
import { toYMap } from '../lib/toYMap.js'
import { group, ungroup } from '../lib/group.js'
import { newsDocToSlate, slateToNewsDoc } from '../newsdoc/index.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import { type TBElement } from '@ttab/textbit'
import { type Document } from '@hocuspocus/server'
import createHash from '../../../../shared/createHash.js'

function yContentToNewsDoc(yContent: Y.XmlText): Block[] | undefined {
  const slateElement = yTextToSlateElement(yContent).children
  return slateToNewsDoc(slateElement as TBElement[])
}

function assertDescriptions(yMeta: Y.Map<unknown>, documentType: string): void {
  // Only perform on planning items and assignments
  if (!['core/planning-item', 'core/assignment'].includes(documentType)) {
    return
  }

  const yDesc = yMeta.get('core/description') as Y.Array<Y.Map<unknown>>

  const payload: Block = {
    id: '',
    uuid: '',
    uri: '',
    url: '',
    type: 'core/description',
    title: '',
    data: { text: '' },
    rel: '',
    role: '',
    name: '',
    value: '',
    contentType: '',
    links: [],
    content: [],
    meta: []
  }

  // No descriptions
  if (!yDesc) {
    const yArr = new Y.Array()
    yArr.push([toYMap({ ...payload, role: 'public' }, new Y.Map())])
    yArr.push([toYMap({ ...payload, role: 'internal' }, new Y.Map())])

    yMeta.set('core/description', yArr)
    return
  }

  // Already has two descriptions
  if (yDesc.length === 2) {
    return
  }

  // Find missing description
  const descriptionTypes = yDesc.map((yMap) => yMap.get('role') as string)
  const missingType = ['public', 'internal'].filter((role) => !descriptionTypes.includes(role))

  missingType.forEach((role) => {
    yDesc.push([toYMap({ ...payload, role }, new Y.Map())])
  })
}

export function newsDocToYDoc(yDoc: Document | Y.Doc, newsDoc: GetDocumentResponse): void {
  try {
    const yMap = yDoc.getMap('ele')
    const { document, version } = newsDoc

    if (document) {
      const { meta, links, content, ...rest } = document

      yMap.set('meta', toYMap(group(document.meta, 'type'), new Y.Map()))
      yMap.set('links', toYMap(group(document.links, 'type'), new Y.Map()))
      yMap.set('root', toYMap(rest, new Y.Map()))

      assertDescriptions(yMap.get('meta') as Y.Map<unknown>, document.type)

      // Assert assignment descriptions
      const yMeta = yMap.get('meta') as Y.Map<unknown>
      const yAssignments = yMeta.get('core/assignment') as Y.Array<unknown>
      if (yAssignments?.length) {
        for (const yAssignment of yAssignments) {
          const assMeta = (yAssignment as Y.Map<unknown>).get('meta') as Y.Map<unknown>
          assertDescriptions(assMeta, 'core/assignment')
        }
      }

      // Share editable content for Textbit use
      const yContent = new Y.XmlText()
      const slateDocument = newsDocToSlate(document?.content ?? [])
      yContent.applyDelta(
        slateNodesToInsertDelta(slateDocument)
      )
      yMap.set('content', yContent)

      // Set version
      const yVersion = yDoc.getMap('version')
      yVersion?.set('version', version?.toString())

      const originalHash = createHash(JSON.stringify(newsDoc.document))

      const yOriginalHash = yDoc.getMap('hash')
      yOriginalHash?.set('hash', originalHash)

      return
    }

    throw new Error('No document')
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}

export function yDocToNewsDoc(yDoc: Y.Doc): GetDocumentResponse {
  const yMap = yDoc.getMap('ele')
  try {
    const meta = ungroup((yMap.get('meta') as Y.Map<unknown>)?.toJSON() || {})
    const links = ungroup((yMap.get('links') as Y.Map<unknown>)?.toJSON() || {})

    const yContent = yMap.get('content') as Y.XmlText
    const content = yContent.toString() ? yContentToNewsDoc(yContent) : []


    const root = yMap.get('root') as Y.Map<unknown>

    const { uuid, type, uri, url, title, language } = root.toJSON()

    meta.forEach(docMeta => {
      if (docMeta.type === 'core/assignment') {
        docMeta.meta = docMeta.meta.filter(assMeta => {
          if (assMeta.type === 'core/description') {
            return assMeta.data.text !== ''
          }
          return true
        })
      }
    })

    return {
      version: BigInt(yDoc.getMap('version').get('version') as string),
      document: {
        uuid,
        type,
        uri,
        url,
        title,
        content: content || [],
        // Remove added empty descriptions
        meta: meta.filter((m) => {
          if (m.type === 'core/description') {
            return m.data.text !== ''
          }
          return true
        }),
        links,
        language
      }
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}
