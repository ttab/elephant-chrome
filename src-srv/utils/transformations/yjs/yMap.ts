import * as Y from 'yjs'
import { type Block, type GetDocumentResponse } from '../../../protos/service.js'
import { toYMap } from '../lib/toYMap.js'
import { group, ungroup } from '../lib/group.js'
import { newsDocToSlate, slateToNewsDoc } from '../newsdoc/index.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import { type TBElement } from '@ttab/textbit'
import { type Document } from '@hocuspocus/server'

function yContentToNewsDoc(yContent: Y.XmlText): Block[] | undefined {
  const slateElement = yTextToSlateElement(yContent).children
  return slateToNewsDoc(slateElement as TBElement[])
}

export function newsDocToYMap(yDoc: Document | Y.Doc, newsDoc: GetDocumentResponse): void {
  try {
    const yMap = yDoc.getMap('ele')
    const { document, version } = newsDoc

    if (document) {
      const { meta, links, content, ...rest } = document

      yMap.set('meta', toYMap(group(document.meta, 'type'), new Y.Map()))
      yMap.set('links', toYMap(group(document.links, 'type'), new Y.Map()))
      yMap.set('root', toYMap(rest, new Y.Map()))

      // Share editable content for Textbit use
      const yContent = new Y.XmlText()
      const slateDocument = newsDocToSlate(document?.content ?? [])
      yContent.applyDelta(
        slateNodesToInsertDelta(slateDocument)
      )
      yMap.set('content', yContent)

      // Set version
      const yVersion = yDoc.getMap('version')
      yVersion?.set('version', version?.toString())

      return
    }

    throw new Error('No document')
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}

export function yDocToNewsDoc(yDoc: Y.Doc): GetDocumentResponse {
  const yMap = yDoc.getMap('ele')
  try {
    const meta = ungroup((yMap.get('meta') as Y.Map<unknown>)?.toJSON() || {})
    const links = ungroup((yMap.get('links') as Y.Map<unknown>)?.toJSON() || {})

    const yContent = yMap.get('content') as Y.XmlText
    const content = yContent.toString() ? yContentToNewsDoc(yContent) : []


    const root = yMap.get('root') as Y.Map<unknown>

    const { uuid, type, url, uri, title, language } = root.toJSON()

    return {
      version: BigInt(yDoc.getMap('version').get('version') as string),
      document: {
        uuid,
        type,
        url,
        uri,
        title,
        content: content || [],
        meta,
        links,
        language
      }
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}
