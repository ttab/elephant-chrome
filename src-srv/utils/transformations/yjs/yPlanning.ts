import * as Y from 'yjs'
import { type Document } from '../../../protos/service.js'
import { toYMap } from '../lib/toYMap.js'
import { group, ungroup } from '../lib/group.js'

export function newsDocToYPlanning(document: Document, planningYMap: Y.Map<unknown>): Y.Map<unknown> {
  try {
    const { meta, links, ...rest } = document

    planningYMap.set('meta', toYMap(group(document.meta, 'type'), new Y.Map()))
    planningYMap.set('links', toYMap(group(document.links, 'type'), new Y.Map()))
    planningYMap.set('root', toYMap(rest, new Y.Map()))

    return planningYMap
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}

export function yPlanningToNewsDoc(planningYMap: Y.Map<unknown>): Document {
  try {
    const meta = ungroup((planningYMap.get('meta') as Y.Map<unknown>)?.toJSON() || {})
    const links = ungroup((planningYMap.get('links') as Y.Map<unknown>)?.toJSON() || {})
    const content = ungroup((planningYMap.get('content') as Y.Map<unknown>)?.toJSON() || {})
    const root = planningYMap.get('root') as Y.Map<unknown>

    const res = {
      ...root.toJSON(),
      content,
      meta,
      links
    } as unknown as Document

    return res
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}

