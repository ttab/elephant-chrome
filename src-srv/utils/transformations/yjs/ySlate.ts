import * as Y from 'yjs'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import type { GetDocumentResponse } from '../../../protos/service.js'
import { type TBElement } from '@ttab/textbit'
import { slateToNewsDoc } from '../index.js'

export function yjsStateAsUpdate(content: TBElement[], data: Y.Doc): Uint8Array {
  const insertContentDelta = slateNodesToInsertDelta(content)
  const sharedContentRoot = data.get('content', Y.XmlText) as Y.XmlText
  sharedContentRoot.applyDelta(insertContentDelta)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-argument
  data.share.set('content', sharedContentRoot as Y.AbstractType<any>)

  return Y.encodeStateAsUpdate(data)
}

/**
* Helper function to transform a Document(Y.Doc) to a newsDoc
* @param document Document
* @returns GetDocumentResponse
*/
export function yDocToNewsDoc(document: Y.Doc): GetDocumentResponse {
  const original = document.get('original', Y.Map) as Y.Map<unknown>
  const json = original.toJSON()

  // revert content
  const sharedRoot = document.get('content', Y.XmlText) as Y.XmlText
  const content: TBElement[] = yTextToSlateElement(sharedRoot).children
  const version: number = json.version

  json.document.content = slateToNewsDoc(content)
  json.version = BigInt(version)

  return json as GetDocumentResponse
}
