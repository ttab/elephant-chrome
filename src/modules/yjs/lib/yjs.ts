import * as Y from 'yjs'
import type { EleDocumentResponse } from '@/shared/types'
import { slateNodesToInsertDelta } from '@slate-yjs/core'

export type YPath = [string, ...(string | number)[]]

type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, unknown>
    ? DeepPartial<T[P]> // Recursively apply partial updates
    : T[P]
}

/*
 * Create a typed YJS document from a given object.
 */
export function createTypedYDoc(
  data?: EleDocumentResponse,
  options?: {
    document?: Y.Doc
    rootMap?: string
    metaMap?: string
    isReady?: boolean
  }
): Y.Doc {
  // Base document and base maps
  const ydoc = options?.document ?? new Y.Doc()
  const ymap = ydoc.getMap(options?.rootMap ?? 'document')
  const yMeta = ydoc.getMap(options?.metaMap ?? '__meta')

  // Setup __meta (system only information about the document)
  if (options && Object.hasOwnProperty.call(options, 'isReady')) {
    yMeta.set('isReady', options.isReady)
  }

  if (!data?.document) {
    return ydoc
  }

  function populateYMap(obj: EleDocumentResponse, ymap: Y.Map<unknown>) {
    for (const [key, value] of Object.entries(obj)) {
      if (Array.isArray(value)) {
        const yarray = new Y.Array<unknown>()
        value.forEach((item) => {
          if (typeof item === 'object' && item !== null) {
            const childMap = new Y.Map<unknown>()
            populateYMap(item as EleDocumentResponse, childMap)
            yarray.push([childMap])
          } else {
            yarray.push([item])
          }
        })
        ymap.set(key, yarray)
      } else if (typeof value === 'object' && value !== null) {
        const childMap = new Y.Map<unknown>()
        populateYMap(value as EleDocumentResponse, childMap)
        ymap.set(key, childMap)
      } else {
        ymap.set(key, value)
      }
    }
  }

  populateYMap(data, ymap)

  // EleDocument specific handling
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(data.document.content)
  )

  ymap.set('content', yContent)

  // Set version
  const yVersion = ydoc.getMap('version')
  yVersion?.set('version', data.version)

  return ydoc
}

/*
 * Recursively extracts data from a YJS structure
 */
export function extractYData<T>(y: Y.Map<unknown> | Y.Array<unknown> | Y.XmlText): T {
  if (y instanceof Y.XmlText) {
    return y.toString() as T
  }

  if (y instanceof Y.Map) {
    const result: Record<string, unknown> = {}

    for (const [key, value] of y.entries()) {
      if (value instanceof Y.Map || value instanceof Y.Array) {
        result[key] = extractYData(value)
      } else {
        result[key] = value
      }
    }

    return result as T
  }

  if (y instanceof Y.Array) {
    return y.toArray().map((item) =>
      item instanceof Y.Map || item instanceof Y.Array
        ? extractYData(item)
        : item
    ) as T
  }

  throw new Error('Unsupported Yjs type')
}

/*
 * Recursively updates a YJS document
 *
 * @example
 * updateYMap(document.get('document'), { meta: { priority: Number(value) } })
 */
export function updateYMap<T extends Record<string, unknown>>(ymap: Y.Map<unknown>, newData: DeepPartial<T>) {
  ymap.doc?.transact(() => { // Ensure all changes are atomic
    for (const [key, value] of Object.entries(newData)) {
      if (Array.isArray(value)) {
        const yarray = new Y.Array<unknown>()
        value.forEach((item) => {
          if (typeof item === 'object' && item !== null) {
            const childMap = new Y.Map<unknown>()
            updateYMap(childMap, item as Record<string, unknown>)
            yarray.push([childMap])
          } else {
            yarray.push([item])
          }
        })
        ymap.set(key, yarray)
      } else if (typeof value === 'object' && value !== null) {
        let existingYMap = ymap.get(key) as Y.Map<unknown> | undefined
        if (!(existingYMap instanceof Y.Map)) {
          existingYMap = new Y.Map<unknown>()
          ymap.set(key, existingYMap)
        }
        updateYMap(existingYMap, value as Record<string, unknown>)
      } else {
        ymap.set(key, value)
      }
    }
  })
}

/**
 * Converts a string path to the same array path format used by yjs maintaing
 * the difference between strings for map properties and numbers for array positions.
 *
 * Example:
 * 'meta.myArray[1].value' -> ['meta', 'myArray', 1, 'value]
 */
export function stringToYPath(input: string): YPath {
  const result: Partial<YPath> = []
  const regex = /([^[\].]+)|\[(\d+)\]/g
  let match: RegExpExecArray | null

  while ((match = regex.exec(input)) !== null) {
    if (match[1]) {
      // Matched a word (property in a Y.Map)
      result.push(match[1])
    } else if (match[2]) {
      // Matched a number in brackets (array index)
      result.push(parseInt(match[2]))
    }
  }

  return result as YPath
}

type YjsContainer = Y.Map<unknown> | Y.Array<unknown> | Y.Text | Y.XmlText | Y.XmlFragment | Y.XmlElement

/**
 * Get the path of a Yjs value within its parent structure.
 *
 * @param {Y.AbstractType<unknown>} v - The Yjs value to get the path for.
 * @param {(string | number)[]} path - The current path being constructed.
 * @returns {(string | number)[]} The path of the Yjs value as an array of strings and numbers.
 */
export function getYjsPath(value: YjsContainer | undefined, asString: true): string
export function getYjsPath(value: YjsContainer | undefined, asString?: false): (string | number)[]
export function getYjsPath(value: YjsContainer | undefined, asString: boolean): string | (string | number)[]
export function getYjsPath(
  v?: YjsContainer,
  asString: boolean = false
): (string | number)[] | string {
  const path: (string | number)[] = []

  let current = v
  while (current?._item) {
    const { parent, parentSub } = current._item

    if (parent instanceof Y.Map) {
      path.unshift(parentSub as string)
      current = parent as YjsContainer
    } else if (parent instanceof Y.Array) {
      const index = parent.toArray().indexOf(current)
      path.unshift(index)
      current = parent as YjsContainer
    } else {
      break
    }
  }

  return asString ? path.join('.') : path
}
