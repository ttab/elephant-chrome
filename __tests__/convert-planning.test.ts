import { newsDocToYMap, yMapToNewsDoc } from '../src-srv/utils/transformations/yjs/yMap'

import * as Y from 'yjs'

import { planning as pl } from './data/planning-repo'

describe('Conversion of newsdoc to yMap document', () => {
  const { document } = pl
  const yDoc = new Y.Doc()
  const yMap = yDoc.getMap('ele')

  if (!document) {
    throw new Error('no document')
  }

  const yPlanning = newsDocToYMap(document, yMap)

  it('transforms newsDoc to yPlanning', () => {
    const planningJson = yDoc.getMap('ele').toJSON()
    expect(planningJson).toMatchSnapshot()
  })

  it('transforms yPlanning to newsDoc', () => {
    const newsDoc = yMapToNewsDoc(yPlanning)
    expect(newsDoc).toMatchSnapshot()
  })
})
