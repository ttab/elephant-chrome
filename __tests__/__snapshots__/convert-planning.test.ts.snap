// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversion of newsdoc to yMap document transforms newsDoc to yPlanning 1`] = `
Object {
  "links": Object {
    "tt/sector": Array [
      Object {
        "content": Object {},
        "contentType": "",
        "data": Object {},
        "id": "",
        "links": Object {},
        "meta": Object {},
        "name": "",
        "rel": "sector",
        "role": "",
        "title": "Sport",
        "type": "tt/sector",
        "uri": "sector://spt",
        "url": "",
        "uuid": "a36ff853-9fb3-5950-8893-64ac699f5481",
        "value": "SPT",
      },
    ],
  },
  "meta": Object {
    "core/assignment": Array [
      Object {
        "content": Object {},
        "contentType": "",
        "data": Object {
          "end": "2024-02-09T22:59:59Z",
          "end_date": "2024-02-09",
          "full_day": "true",
          "public": "true",
          "publish": "2024-02-09T10:30:00Z",
          "start": "2024-02-08T23:00:00Z",
          "start_date": "2024-02-09",
        },
        "id": "97464374-3d0c-44b9-9126-f5efadbd32a7",
        "links": Object {
          "core/article": Array [
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "f283c9a0-6a2e-4021-a009-087961dd032f",
              "value": "",
            },
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "fa5e6fbf-c03d-4447-8a1f-550176979a48",
              "value": "",
            },
          ],
          "core/author": Array [
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "John Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "c37fdf3e-72ff-4e22-8b9f-1af0d60b0cd9",
              "value": "",
            },
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "Jane Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "f23c55de-5e52-428e-9ed7-0cdc0921f60a",
              "value": "",
            },
          ],
        },
        "meta": Object {
          "core/assignment-type": Array [
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "core/assignment-type",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "text",
            },
          ],
          "tt/slugline": Array [
            Object {
              "content": Object {},
              "contentType": "",
              "data": Object {},
              "id": "",
              "links": Object {},
              "meta": Object {},
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "tt/slugline",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "lands-tomasson",
            },
          ],
        },
        "name": "",
        "rel": "",
        "role": "",
        "title": "Kortare text utsänd. Uppdateras",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "core/planning-item": Array [
      Object {
        "content": Object {},
        "contentType": "",
        "data": Object {
          "end_date": "2024-02-09",
          "priority": "4",
          "public": "true",
          "start_date": "2024-02-09",
          "tentative": "false",
        },
        "id": "",
        "links": Object {},
        "meta": Object {},
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
  },
  "root": Object {
    "content": Array [],
    "language": "sv-se",
    "title": "Tomasson lämnar Blackburn – för landslaget?",
    "type": "core/planning-item",
    "uri": "core://newscoverage/726045e7-a52b-4737-8fba-c8149f7e2c2d",
    "url": "",
    "uuid": "726045e7-a52b-4737-8fba-c8149f7e2c2d",
  },
}
`;

exports[`Conversion of newsdoc to yMap document transforms yPlanning to newsDoc 1`] = `
Object {
  "content": Array [],
  "language": "sv-se",
  "links": Array [
    Object {
      "content": Array [],
      "contentType": "",
      "data": Object {},
      "id": "",
      "links": Array [],
      "meta": Array [],
      "name": "",
      "rel": "sector",
      "role": "",
      "title": "Sport",
      "type": "tt/sector",
      "uri": "sector://spt",
      "url": "",
      "uuid": "a36ff853-9fb3-5950-8893-64ac699f5481",
      "value": "SPT",
    },
  ],
  "meta": Array [
    Object {
      "content": Array [],
      "contentType": "",
      "data": Object {
        "end_date": "2024-02-09",
        "priority": "4",
        "public": "true",
        "start_date": "2024-02-09",
        "tentative": "false",
      },
      "id": "",
      "links": Array [],
      "meta": Array [],
      "name": "",
      "rel": "",
      "role": "",
      "title": "",
      "type": "core/planning-item",
      "uri": "",
      "url": "",
      "uuid": "",
      "value": "",
    },
    Object {
      "content": Array [],
      "contentType": "",
      "data": Object {
        "end": "2024-02-09T22:59:59Z",
        "end_date": "2024-02-09",
        "full_day": "true",
        "public": "true",
        "publish": "2024-02-09T10:30:00Z",
        "start": "2024-02-08T23:00:00Z",
        "start_date": "2024-02-09",
      },
      "id": "97464374-3d0c-44b9-9126-f5efadbd32a7",
      "links": Array [
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "John Doe/TT",
          "rel": "assignee",
          "role": "primary",
          "title": "",
          "type": "core/author",
          "uri": "",
          "url": "",
          "uuid": "c37fdf3e-72ff-4e22-8b9f-1af0d60b0cd9",
          "value": "",
        },
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "Jane Doe/TT",
          "rel": "assignee",
          "role": "primary",
          "title": "",
          "type": "core/author",
          "uri": "",
          "url": "",
          "uuid": "f23c55de-5e52-428e-9ed7-0cdc0921f60a",
          "value": "",
        },
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "",
          "rel": "deliverable",
          "role": "",
          "title": "",
          "type": "core/article",
          "uri": "",
          "url": "",
          "uuid": "f283c9a0-6a2e-4021-a009-087961dd032f",
          "value": "",
        },
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "",
          "rel": "deliverable",
          "role": "",
          "title": "",
          "type": "core/article",
          "uri": "",
          "url": "",
          "uuid": "fa5e6fbf-c03d-4447-8a1f-550176979a48",
          "value": "",
        },
      ],
      "meta": Array [
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "",
          "rel": "",
          "role": "",
          "title": "",
          "type": "tt/slugline",
          "uri": "",
          "url": "",
          "uuid": "",
          "value": "lands-tomasson",
        },
        Object {
          "content": Array [],
          "contentType": "",
          "data": Object {},
          "id": "",
          "links": Array [],
          "meta": Array [],
          "name": "",
          "rel": "",
          "role": "",
          "title": "",
          "type": "core/assignment-type",
          "uri": "",
          "url": "",
          "uuid": "",
          "value": "text",
        },
      ],
      "name": "",
      "rel": "",
      "role": "",
      "title": "Kortare text utsänd. Uppdateras",
      "type": "core/assignment",
      "uri": "",
      "url": "",
      "uuid": "",
      "value": "",
    },
  ],
  "title": "Tomasson lämnar Blackburn – för landslaget?",
  "type": "core/planning-item",
  "uri": "core://newscoverage/726045e7-a52b-4737-8fba-c8149f7e2c2d",
  "url": "",
  "uuid": "726045e7-a52b-4737-8fba-c8149f7e2c2d",
}
`;
